src/UglyToad.PdfPig/Geometry/PdfPoint.cs

```diff
@@ -82,6 +82,39 @@ internal PdfVector ToVector()
             return new PdfVector(X, Y);
         }
 
+        /// <summary>
+        /// Converts this <see cref="PdfPoint"/> into an array.
+        /// </summary>
+        /// <returns></returns>
+        public double[] ToDouble()
+        {
+            return new double[] { (double)this.X, (double)this.Y };
+        }
+
+        /// <summary>
+        /// Returns a value indicating whether this <see cref="PdfPoint"/> is equal to a specified <see cref="PdfPoint"/> .
+        /// </summary>
+        /// <param name="obj"></param>
+        /// <returns></returns>
+        public override bool Equals(object obj)
+        {
+            if (obj is PdfPoint)
+            {
+                PdfPoint point = (PdfPoint)obj;
+                return point.X == this.X && point.Y == this.Y;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Returns the hash code for this <see cref="PdfPoint"/>.
+        /// </summary>
+        /// <returns></returns>
+        public override int GetHashCode()
+        {
+            return (int)(this.X * 10_000 + 31 * this.Y * 10_000);
```

Comment:
Overriding `GetHashCode` is one of those annoying things that we're told to do but correct approaches to it aren't really documented anywhere. There's a couple of discussions on it:

+ https://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-an-overridden-system-object-gethashcode/263416#263416
+ https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/

Initially I thought the approach here might be more vulnerable to collisions but now I'm not so sure, though I guess the only way to be sure is calculate the collision rate or something. I tend to treat hash codes as something like encryption in that they're much easier to use library functions for and easy to get wrong when written from scratch.

I generally rely on Resharper when doing this which generates something that closely matches Jon Skeet's answer. For reference this looks something like:

```
public override int GetHashCode()
{
    var hashCode = 1861411795;
    hashCode = hashCode * -1521134295 + X.GetHashCode();
    hashCode = hashCode * -1521134295 + Y.GetHashCode();
    return hashCode;
}
```

The other option is to let the .NET Framework do the heavy lifting by relying on the `ValueTuple` implementation:

```return (X, Y).GetHashCode();```

Which turns the point into a ValueTuple and uses the implementation from that which is here: https://github.com/dotnet/roslyn/blob/master/src/Compilers/Test/Resources/Core/NetFX/ValueTuple/ValueTuple.cs#L297. Now I'm not sure what that does but I'd hope it's one of the best possible implementations.

Reply:
Hehe exactly. Let's .Net do the job then!